crates/router/src/core/payments/helpers.rs

```diff
@@ -413,75 +417,127 @@ pub async fn get_token_pm_type_mandate_details(
+        payment_method,
+        payment_method_type,
+        mandate_data,
+        recurring_payment_data,
+        mandate_connector_details,
+        payment_method_info,
+    ) = match mandate_type {
+        Some(api::MandateTransactionType::NewMandateTransaction) => (
+            request.payment_token.to_owned(),
+            request.payment_method,
+            request.payment_method_type,
+            Some(mandate_data.clone().get_required_value("mandate_data")?),
+            None,
+            None,
+            None,
+        ),
         Some(api::MandateTransactionType::RecurringMandateTransaction) => {
-            let (
-                token_,
-                payment_method_,
-                recurring_mandate_payment_data,
-                payment_method_type_,
-                mandate_connector,
-            ) = get_token_for_recurring_mandate(
-                state,
-                request,
-                merchant_account,
-                merchant_key_store,
-            )
-            .await?;
-            Ok((
-                token_,
-                payment_method_,
-                payment_method_type_.or(request.payment_method_type),
-                None,
-                recurring_mandate_payment_data,
-                mandate_connector,
-            ))
+            match &request.recurring_details {
+                Some(recurring_details) => match recurring_details {
+                    RecurringDetails::MandateId(mandate_id) => {
+                        let mandate_generic_data = get_token_for_recurring_mandate(
+                            state,
+                            request,
+                            merchant_account,
+                            merchant_key_store,
+                            mandate_id.to_owned(),
+                        )
+                        .await?;
+
+                        (
+                            mandate_generic_data.token,
+                            mandate_generic_data.payment_method,
+                            mandate_generic_data
+                                .payment_method_type
+                                .or(request.payment_method_type),
+                            None,
+                            mandate_generic_data.recurring_mandate_payment_data,
+                            mandate_generic_data.mandate_connector,
+                            None,
+                        )
+                    }
+                    RecurringDetails::PaymentMethodId(payment_method_id) => {
+                        let payment_method_info = state
+                            .store
+                            .find_payment_method(payment_method_id)
+                            .await
+                            .to_not_found_response(
+                                errors::ApiErrorResponse::PaymentMethodNotFound,
+                            )?;
+
+                        (
+                            request.payment_token.to_owned(),
```

Comment:
If recurring details are provided, then token is not required, so we can hardcode this to `None`. It would be better to add validation for this as well in the PaymentsConfirm `ValidateRequest` operation (both token and recurring details cannot be given).

```suggestion
                            request.payment_token.to_owned(),
```